rfcs/extended_goal_state_functionality.md

```diff
@@ -0,0 +1,72 @@
+
+# Long-Term Objectives
+
+Following discussions prior to this RFC, we have identified three objectives related to the functional and architectural changes proposed here.
+
+  1. Implement the functionality of `GetGoodView` for the distant agent within the proposed framework. Goal-state-generating SMs will provide the model-free signals used to orient the agent to on-object regions. These signals can be filtered out or overridden when learning modules emit goal states, and the filtering will be performed by a `GoalStateSelector`. Note this `GetGoodView` will be an inherent part of Monty, and can be called, for example, after a `JumpToGoalState` causes Monty to lose the object. This is thus distinct from the version of `GetGoodView` being extracted as a `PositioningProcedure`. Positioning Procedures cannot be called by Monty, and happen outside of the inference or learning episode. 
+  2. Integration of model-free segmentation and model-based policies to recognize an object before moving on to a new object. This objective is the primary motive that led to the formation of this specific RFC.
+  3. Improve efficiency by learning and using sparser models. It has been suggested that our learned models are unnecessarily dense owing to the fact that non-salient and salient parts of an object are represented equally. If we can ensure that only the most salient parts of an object are attended to during inference, then we can effectively ignore (or at least sample less frequently) the non-salient areas during training without degrading performance.
+
+Each of these tasks is significant, likely warranting its own RFC. Further discussion is outside the scope of this RFC.
+
+# Discussion
+
+### Dataloader
+The changes proposed here may disrupt the existing goal-state-execution system. More specifically, the dataloader is currently responsible for checking whether a motor system has a valid goal state. If so, it attempts to satisfy the goal state by "jumping" to a new location. In the proposed system, goal states will no longer be coupled to a single execution method, and all conversions between goal states to actions will be performed within the motor system. Consequently, the dataloader's "jump-to-goal-state" code will need to live elsewhere, preferably somewhere it can be reused by different polices. Fortunately, removing any reliance on the dataloader to execute goal states is consistent with existing plans.
+
+### Computing Goal-State Success
+We are still determining the best way to verify when a goal state has been achieved. As it stands, each `GoalStateGenerator` performs this task by reaching into its parent LM, grabbing its input observations, and comparing them with the previously generated goal state. Setting aside parent-access, reimplementing this pattern for goal-state-generating sensor modules won't work for a couple of reasons:
```

Comment:
What would this information be used for? Are you thinking of eventually having GSGs that learn by themselves what goalstates to output and use feedback about success to adjust this? It seems like that would be nice in the future, but for the current GSGs, we are planning on it is not necessary.

Reply:
Yeah I was questioning what this information would be used for? Like how would it change the goal state generator's output if the goal state was achieved or not achieved? It seems like unless the GSG is actively learning and updating it's policy based on feedback (which we are not planning on for these first model-free policies) it would have no effect.

Reply:
Yeah I think that's an interesting question. In the long-term it could be useful for a GSG to explicitly check for whether a goal-state is successful or not after it outputs a goal state, based on the next sensory inputs. This would enable it to learn what goal-states are correlated with changes in the world. I agree with you that with the current setup, just checking the conditions for whether a goal-state should be generated would be functionality equivalent. 

For what it's worth, we already have a `_check_output_goal_state_achieved` method, so this wouldn't be adding anything.